### ISSUE TITLE
Membership protocol

### ISSUE THREAD

**Issue by dirkmc** [peer-base/peer-base#243]
Currently we use App level gossip to propagate collaboration membership changes. When a peer detects that a member was added or removed from a collaboration, it broadcasts the CRDT state to all other peers. Each peer broadcasts that state again, and so forth, until the only messages received by any peer reflect the unanimously agreed upon state. Therefore each membership change triggers O(N²) messages, where N is the number of peers participating in the App.

App level gossip is implemented over floodsub:
- On startup the local peer connects to a rendezvous server and listens for `peer` events
- When a `peer` event is received from rendezvous, the local peer dials the remote peer
- When a connection is made, Floodsub receives a `peer:connect` event and dials the peer over the `floodsub` protocol
- If the remote peer is listening on the `floodsub` protocol, Floodsub sends the local peer's subscriptions to the remote peer
- When the local peer receives subscription information from a remote peer, it checks if the remote peer is subscribed to the App name. If so the local peer adds the remote peer to the ring of App peers
- The local peer recalculates the Dias set from the ring, and connects to all peers in the Dias set and disconnects from peers that are not in the Dias set
- When the local peer broadcasts a gossip message, it sends the message to all peers it is connected directly to (the peers in the Dias set), which forward the message on to their direct connections, and so forth
- Each message has a message ID. If the peer receives a message with a message ID that it has seen recently, it ignores the message

I propose that we replace floodsub with a new membership protocol:
- On startup the local peer connects to a rendezvous server and listens for `peer` events
- When a `peer` event is received from rendezvous, the local peer dials the remote peer over the `membership` protocol, and adds the peer to the App peers ring
- The local peer recalculates the Dias set from the ring, and connects to all peers in the Dias set and disconnects from peers that are not in the Dias set
- When the local peer detects a membership change, it sends the membership CRDT to its neighbouring peers
- Those peers forward the message on to their neighbours, and keep track of the neighbours' vector clocks, so that only messages with new information are sent to a neighbour

This change would
- simplify peer discovery: no need to wait for subscription information to be transferred amongst peers by floodsub
- speed up peer discovery: Perform a single `membership` protocol dial, instead of
  - a regular dial
  - then a `floodsub` protocol dial
  - then subscription information exchange
- reduce the number of messages sent on a membership change from O(N²) to O(N)

Note: when rendezvous fires a `peer` event with the associated `PeerInfo`, it should be possible to examine the `PeerInfo.protocols` field to decide whether to bother dialing that peer. However currently rendezvous does not support the `PeerInfo.protocols` field, so it might be worth asking for it to be supported.

**Comment by pgte** [comment]
@dirkmc I think this would work well.
Ideally, we wouldn't need to send the entire state of the CRDT, but only the deltas as they occur.

**Comment by dirkmc** [comment]
Right, it would be similar to how the push/pull protocols work for collaborations 👍

**Comment by pgte** [comment]
@dirkmc perhaps we could just use the same stack as the collaboration data for syncing membership?

**Comment by dirkmc** [comment]
The use case is similar, but simpler. In comparison to collaboration state updates, membership changes are both less frequent and more urgent, so probably we don't need lazy / eager mode (effectively it would always be in eager mode) and there is no special behaviour for pinners. We can probably reuse some of the code.

**Comment by dirkmc** [comment]
I think that makes sense, there would be some refactoring but the protocols would be essentially the same

**Comment by dirkmc** [comment]
I'm thinking of implementing the protocol such that we can keep track of members of the app and of each collaboration using a two-level CRDT:
1. an app-wide CRDT of `<peer id> => [peer addresses]` as an `ormap => mvreg`
2. a CRDT with peer ids for each member of the collaboration as an `aworset`

The CRDTs would be synchronized using an app-wide collaboration (1) with sub-collaborations for each collaboration membership (2)

eg
```
  my app: {
    <peer A> => [ <address 1>, <address 2, ... ],
    <peer B> => [ <address 1>, <address 2, ... ]
  }

 my collab 1: [ <peer A>, <peer B> ]
 my collab 2: [ <peer A>, <peer C> ]
```

With this approach there would now be two sources of discovery:
1. The rendezvous service
2. The app-wide CRDT

@pgte do you see any issues with having two sources of discovery?
One potential issue I see is where each peer is able to connect to a different subset of peers because of firewalls etc
1. Peer A successfully dials Peer C and adds it to the app-wide CRDT
2. Peer B receives the CRDT update, and dials Peer C but is unable to connect, so it removes Peer B from the CRDT
3. Repeat step 1

I guess this is already an issue with the existing implementation in the sense that if members of a collaboration have distinct connectivity to a subset of peers, they will endlessly gossip membership state while trying to converge.

Another question I have is whether it would be abusing the concept of libp2p protocols to have a different protocol name for each app, eg "/peer-base/my-app/membership", "/peer-base/another-app/membership" etc?

**Comment by pgte** [comment]
Before that, I'm worried that an app-wide CRDT won't scale very well.
IMO, having all the application peer addresses in memory may not be feasible.
Also, the amount of concurrent changes in that CRDT will generate a lot of network activity.

@dirkmc Could we think of a way to avoid sharing app-global state?

**Comment by dirkmc** [comment]
That's true, there are some scalability concerns with this approach. What should the scalability limits be for a peer base app / collaboration?
In order to form a Dias set from the ring, the app needs to know about all the other peers in the ring. If we want to scale beyond a certain point we would probably need to use a different approach, eg a DHT

**Comment by pgte** [comment]
Sorry about being unresponsive on this. I have to get me some more coffee and think about this one for a bit more.. :)

**Comment by dirkmc** [comment]
I'm also thinking about it a little more :)

Stepping back a bit, the end goal is for members of a collaboration to exchange state as quickly and with as few messages as possible. Nice to haves:
- scale to large numbers of nodes
- small number of nodes can connect with a minimum of overhead
  - establish membership connectivity without expensive DHT-style queries
  - keep handshaking to a minimum
  - avoid connection churn
- small memory footprint
  - only keep track of a subset of all peers in collaboration / app
- minimize number of messages sent
  - keep track of neighbours' state
- limit number of open connections
  - both app-level and collaboration-level connections
  - both inbound and outbound
- connectivity redundancy
 - allow unrelated peer-base apps to relay messages for each other?

Currently we keep track of all peers in the app and collaborations and recalculate the Dias Set and its connections each time any peer goes online or offline. In practice we can ensure full connectivity by keeping track of close predecessors and successors in the ring with high precision. Connections to other points in the ring are optimizations, and thus can be maintained more loosely, allowing us to reduce memory footprint (keeping track of just a subset of the ring) and connection churn.

When a new peer wants to join a collaboration, it can bootstrap by connecting to a subset of peers in the same app (discovered eg through rendezvous). Each existing peer
- merges the new peer's collaboration state
- forwards the collaboration state to any members of the collaboration it is connected to
- responds to the new peer with
  - the merged collaboration state (if the existing peer is in the collaboration)
  - the nearest peers to the new peer that the existing peer knows about
The new peer queries the nearest peers to ask for their nearest peers, and so on, until it finds all the peers in its neighbourhood. It maintains a cap on the number of open connections, optimizing for peers in its neighbourhood that share its collaborations, while also keeping a few connections to far away peers.

This approach allows a new peer to quickly exchange collaboration information, while gradually working towards its place in the ring, and limiting the amount of peer information in memory and the number of open connections. It is near optimal for small numbers of peers and also scales to large numbers of peers.

The Dias Set is similar conceptually to the structure of Chord's DHT, while this approach of asking peers for their neighbours is similar to the Kademlia DHT strategy. The collaboration membership exchange is similar to floodsub topics.

**Comment by aschmahmann** [comment]
@dirkmc @pgte I've been following this a little bit and had a couple questions just to give me some context on the what and why for this proposal.

Are there different requirements here than what PubSub has for discovery? My understanding is that systems on top of PubSub like IPNS over PubSub have a similar discovery issue with trying to figure out how to propagate information to parties interested in **#topic**.

For the App-wide CRDT, if this is just a way of caching peerID information I'd think of it less like a CRDT and more like a best-effort list of peerIDs. As the set of peers using an App increases the CRDT size will increase as well, and the idea of every peer keeping that CRDT (and keeping it updated and consistent) seems unlikely. However, if you ask each peer in an App to keep a list of n peers (some in their collaborations, some not) then a new peer can ask what peers they know and essentially end up with a peer discovery that is potentially benefited by providing a more topic-specific type of random-walk. Additionally, if I understood correctly having the list of addresses in this App-wide cache is not strictly necessary since the DHT could help resolve peerID -> addresses, but if not too expensive would be nice to keep.

**Comment by dirkmc** [comment]
@aschmahmann you're right, finding out which peers are interested in an app/collaboration is essentially the same problem as subscribing to pubsub topics. Currently we use pubsub topics to establish which peers are interested in a particular peer-base app, then exchange collaboration membership information by broadcasting a CRDT over pubsub. The problem is that there are a lot of redundant messages and (as you mentioned) the CRDT can get very big.

The other issue is that as currently implemented, we are equally strict about maintaining all Dias Set connections, where in reality we only need to be strict about immediate neighbours. If we loosen this restriction on far away peers then as you suggest we can maintain smaller lists of peers and random walk (with a heuristic) to find new ones.

It would be nice if we can use pieces of existing DHT implementations to implement these concepts.
